torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py

```diff
@@ -123,3 +123,54 @@ def checkpoint_wrapper(
         )
 
     return CheckpointWrapper(module, checkpoint_impl, offload_to_cpu)
+
+
+def apply_activation_checkpointing_wrapper(
+    model, checkpoint_wrapper_fn=checkpoint_wrapper, check_fn=lambda _, _unused: True
+):
+    """
+    Applies :func:`checkpoint_wrapper` to modules within `model` based on a user-defined
+    configuration. For each module within `model`, the `check_fn` is used to decide
+    whether `module` should be wrapped with :func:`checkpoint_wrapper` or not.
+
+    Note::
+        This function modifies `model` in place and replaces appropriate layers with
+        their checkpoint-wrapped modules.
+    Usage::
+        model = nn.Sequential(
+            nn.Linear(10, 10), nn.Linear(10, 10), nn.Linear(10, 10)
+        )
+        check_fn = lambda m, l: isinstance(l, nn.Linear)
+        apply_activation_checkpointing(model, checkpoint_wrapper_fn=checkpoint_wrapper, check_fn=check_fn)
+    Args:
+        module (nn.Module):
+            The model who's submodules (or self) should be wrapped with activation checkpointing.
+        checkpoint_wrapper_fn (Optional[Callable[nn.Module]])
+            A `Callable` which will wrap modules
+        check_fn (Optional[Callable[nn.Module, nn.Module]])
+            A lambda function which will be passed overall model and current layer and returns
+            ``True`` or ``False`` depending on whether input layer should be wrapped.
+    Returns: None (`model` is modified inplace)
+    """
+    layers = []
+    for name, layer in model.named_modules():
+        if check_fn(model, layer):
+            toks = name.strip().split(".")
+            module = layer
+            layer = model
```

Comment:
Some interesting variable aliasing going on here 😆 

```suggestion
    for name, module in model.named_modules():
        if check_fn(model, module):
            toks = name.strip().split(".")
            layer = model
```

Reply:
Overall, this method is a bit hard to follow for me because the variables keep changing what they mean.

It seems like the point is that we have to find `nn.Module` attributes of an `nn.Module` to make sure to set them as attributes of the new `nn.Module` returned by `checkpoint_wrapper_fn()`?

`checkpoint_wrapper_fn()` is only applied to the original `nn.Module`s of `model` (via `new_layer = checkpoint_wrapper_fn(module)` where `module = layer` and `layer` is from `model.named_modules()`).

Why do we only need to `setattr` for `nn.Module` attributes for the new `nn.Module` returned by `checkpoint_wrapper_fn()`? Why are they different from other non-`nn.Module` attributes?

Reply:
> Why do we only need to setattr for nn.Module attributes for the new nn.Module returned by checkpoint_wrapper_fn()? Why are they different from other non-nn.Module attributes?

We only need to set these since these are new modules, i.e. a `CheckpointWrapper` module containing a submodule which is the wrapped module.

The rest of the modules we didn't wrap are already set? 

Agreed, naming here can definitely be better and function can be made clearer, let me update.